js/src/Ice/ExUtil.js

```diff
@@ -2,19 +2,25 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 //
 
-import { UnexpectedObjectException, MemoryLimitException } from "./LocalExceptions.js";
+import { MarshalException } from "./LocalExceptions.js";
+import { UnknownSlicedValue } from "./UnknownSlicedValue.js";
 
 export function throwUOE(expectedType, v) {
+    // If the object is an unknown sliced object, we didn't find an value factory.
+    if (v instanceof UnknownSlicedValue) {
+        throw new MarshalException(`Cannot find value factory to unmarshal class with type ID '{v.ice_id()}'.`);
+    }
+
     const type = v.ice_id();
-    throw new UnexpectedObjectException(
-        "expected element of type `" + expectedType + "' but received `" + type + "'",
-        type,
-        expectedType,
+    let expected = expectedType.ice_staticId(); // TODO add ice_staticId
```

Comment:
expectedType should be a type ID to start with.

Reply:
Ok, we do the same in C# https://github.com/zeroc-ice/ice/blob/bcdd6a44d862bc788a0764459f79d2d8224a362a/csharp/src/Ice/Internal/Ex.cs#L17